File: sheet/prog/array_list.jl
#!/usr/bin/env julia

# === Array ===
l = [1, 2, 3]	# (3 x 1) Array{Int64,1}
l[1]	# => 1, 1-origin
l[end]	# => 3, 最後の要素
l[1:end]# => 要素すべて (部分配列)
l[:]	# => すべての要素のコピー
view(l, 1:2)	# => 要素1, 2の参照部分配列 (変更できる)
lv = @view l[1:2] # 同上
l2= [1			# ,と同じ
	 2
	 3]

# コピー
l[:]	# => すべての要素のコピー

# 要素数
length(l)	# => 3 : 要素数
size(l)		# => (3,) 

# 検索
findfirst(isequal(2), l)	# =>2, なければnothing
findnext(isequal(2), l, 1)	# pos指定 
# findlast, findprev, findall
# isequal(val), iseven, isodd

# 存在確認
in(1, l)	# => true
1 in l
1 ∈ l


# 要素の追加
# 「!」は対象そのものを変更する関数についている
push!(l, 4, 5)	# l = [1, 2, 3, 4, 5]
println(l)
append!(l, [-1, -2])	# -1, -2を末尾に追加 (Pythonのappend)
println(l)

# 先頭/任意の場所への追加
insert!(l, 1, 10)	# 指定の場所に追加: 1が先頭
pushfirst!(l, 5, 10)	# 5, 10を先頭に追加

prepend!(l, [8, 9])	# appendの先頭バージョン
println(l)
# 削除
println(pop!(l))	# 末尾の要素を取得して削除
println(popfirst!(l))	# 先頭要素
deleteat!(l, 5)		# 5番目の要素を削除

# 
# ソート
println("sort")
println(l)
sort!(l)
println(l)
sort!(l, by = x -> abs(x))	# by = で方法を指定
println(l)

# 走査
println("run")
for i in l    # =, in, ∈も同様に使用可能
    println(i)
end


# 逆順
l = [1, 2, 3, 5, 10]
reverse(l)
reverse!(l)
for i in Iterators.reverse(l)
	println(i)
end

twice = (x:: Int) -> x *2
# map (function applied array)
println("map: ", map(twice, l))
ll = map(twice, l)
# broadcastでもよい
println(twice.(l))
println(l * 2)       # そもそもjuliaならこれでも良い
println(l .* 2)      # そもそもjuliaならこれでも良い
# filter
l3 = filter(isodd, l)	# 奇数のみ取得
println(l3)
# filter!を使用すれば、配列自体を変更できる。
# reduce
prod = reduce(*, l)     # *はjuliaでは関数
println(prod)
# 最大・最小
println(maximum(l))			# => value
println(findmax(l))			# => (value, index)
println(argmax(l))			# => index
println(extrema(l))			# => (min, max)
# minimum, findmin, argmin

# リスト内包表記
println([ n * n for n in l if n > 2 ])
g = ( n * n for n in l if n > 2 ) 	# generator
print(collect(g))	# collectでrangeやgeneratorをArrayに変換

# all
all(isodd, l)
# any
any(isodd, l)


# 多次元配列
m = [1 2 3]		# (1 x 3) Array{Int64,2}
n = [1 2 3		# (3 x 3)
	 4 5 6
	 7 8 9]
n2 = [1 2 3; 4 5 6; 7 8 9]	# 上と同じ
n[1, 1]			# => 1

length(n)	# => 9
size(n)		# => (3, 3)

# 行列の計算
n * n	# 行列の積
n .* n	# アダマール積
log.(n)	# .付き関数はすべての要素に関数を適応する

# 型の指定
Float64[1, 2]		# Floatの配列

# === Tuple ===
# immutable
t = (1, 2, 3, 4, 5)
t[1]	# => 1
(1, )	# 要素数1のTuple
typeof(t)	# => NTuple{5, Int64}

t = ("kcrt", 32)
typeof(t)	# => Tuple{String,Int64}

# 名前付きタプル
t = (a = 1, b = "Hello")
t.a		# => 1
typeof(t)	# => NamedTuple{(:a, :b),Tuple{Int64,String}}

# swap
a = 1; b = 2
a, b = b, a

File: sheet/prog/array_list.py
#!/usr/bin/env python3

# 配列に相当する物は、List, Tuple, Arrayの3種類がある。

from array import array
from functools import reduce
from operator import mul

# === List ===
# 最もよく使用される
l = ['a', 'b', 'c', 'd', 'e', 1, 2, 3]  # 同じ型でなくても良い
l[0]                # => a : 0番目の要素にアクセス
l[-1]               # => 3 : 最後の要素
l[-2]               # => 3 後ろから2番目の要素
l[1:3]              # => ['b', 'c'] : スライス l[1], l[2]からなる新しいリスト
l[3:]               # l[3]から最後まで
l[:]                # 全部のコピー

[]                  # 空のリスト
['A', 'B', 'C'] * 5  # => ['A', 'B', 'C', 'A', 'B' ...]

l2 = l              # l2はlと同じ物をさす。コピーではない！[[注意]]
l2[0] = 'x'         # l2を変更したつもりでも…
l[0]                # => x : lも変更されてしまう！

# 要素のコピー
new_l = l[:]           # スライスか、l.copy()を使って新しいListを得る
new_l[1] = 'z'
l[1]                # => b : これならlは変更されない

# 要素数
len(l)

# 一致の検索
l.count('d')        # 'd'の数を数える

l.index('d')        # => 3 : 最初の'd'の場所、見つからなければ[[ERROR]] ValueError

if 'd' in l:
    pass            # in で存在確認
elif 'd' not in l:
    pass

# 要素の追加: appendとextendがある
a1 = ['A', 'B', 'C']
a2 = ['A', 'B', 'C']
a1.append(['X', 'Y'])   # ['A', 'B', 'C', ['X', 'Y']]
a2.extend(['X', 'Y'])   # ['A', 'B', 'C', 'X', 'Y']

l.append('A')           # 配列でなければappendを使用する
l += ['B', 'C']         # += は extendと同じ


# 先頭/任意の場所への追加
l.insert(0, 'Z')        # 要素を追加 (0で先頭)

# 削除
s = l.pop()             # **最後**の要素を取り出して削除
s = l.pop(0)            # 0 (= 最初)の要素を取り出して削除
l.remove('A')           # 最初の'A'を見つけて消去 ないとValueError[[ERROR]]
del l[3:5]              # 要素3, 4を消去
l.clear()               # 全消去

l = list(range(10))     # [0, 1, 2, ..., 8, 9]
# range()

# ソート: PythonはStableソート
ll = sorted(l)          # ソートされたもの取得
l.sort()                # そのものをソート
l.sort(reverse=True)    # 降順にするにはreverse=True
# カスタムソートは see: https://docs.python.org/ja/3/howto/sorting.html#sortinghowto

# 逆順
ll = reversed(l)
print(ll)               # sortedと違い、イテレータを返す
ll = list(ll)           # これで、通常のリストになる。
l.reverse()             # そのものを逆順に


# 走査
for val in l:
    print(val)

for i, v in enumerate(l):
    print(i, val)

l = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# map (function applied array)
l2 = map(lambda x: x * x, l)
# for v in l2:
#    print(v)
print(l2)
# l2[2]      # [[ERROR]] Mapオブジェクトなので要素でのアクセスは出来ない
l2 = list(l2)       # Listに変換
print(l2[2])


# filter
l2 = filter(lambda x: x > 5, l)
print(l2)
l2 = list(l2)       # Map同様にListに変換

# reduce
print("reduce")
print(l2)
print(reduce(mul, l2, 1))

# 最大・最小
ll = [1, 2, 3, 4]
max(ll)  # => 4
min(ll)  # => 1


# リスト内包表記
# Pythonの強力な機能 [[重要]]
# mapにもfilterにもなる
l2 = [x * x for x in l if x > 5]
# [ 計算式 for 仮引数 in 対象のリスト if 条件 ] (if 条件は省略可)
print(l2)
# []の代わりに () だと generator (必要時に計算)
l3 = (x * x for x in l if x > 5)
print(l3)

# all
ll = [20, 20, 30]
if all(x > 10 for x in ll):
    print("all > 10")

# any
if any(x == 20 for x in ll):
    print("There is 20.")

# === Tuple ===
# イミュータブル(immutable; 変更を許してくれない)なリスト
t = (1, 2, 3, 4, 5)
print(t[4])     # リスト同様に使用できるが、
# t[1] = 10     # [[ERROR]] 変更が必要になる動作は
# t.sort()      # [[ERROR]] 行うことが出来ない
ll = list(t)    # 変更したい場合はリストに変換すれば良い
ll.sort()
t = tuple(ll)   # listからtupleへ

# タプルは代入にも便利
(a, b, c) = (1, 3.2, "Hello")

# === array ===
# 大量の数値を格納するときなどに使う。
# おそらくほとんどのケースで、numpyを使う方がよいので、そちらを参照されたし。
arr = array('i', [1, 2, 3, 4, 5])
arr2 = array('u', "Hello")
# 'b'/'B': singed/unsigned char, 'u': Py_UNICODE
# 'h'/'H': signed/unsigned short, 'i'/'I': signed/unsigned int
# 'l'/'L': signed/unsigned long, 'q'/'Q': signed/unsigned long long
# 'f': float, 'd': double
print(arr.itemsize)     # sizeof(signed int)
arr.append(6), arr.count(3)     # Listと同様にいろいろ使用できる
arr.byteswap()          # リトルエンディアンとビックエンディアンの変換

arr2.fromunicode("Hello")   # こんな名前だが、動作はextend

# array.fromfile(fileobject, n) / array.tofile(fileobject)
